src/scenic/core/vectors.py

```diff
@@ -4,427 +4,659 @@
+
+    def evaluateInner(self, context):
+        return Vector(*(valueInContext(coord, context) for coord in self.coordinates))
+
+    @vectorOperator
+    def applyRotation(self, rotation):
+        if not isinstance(rotation, Orientation):
+            return NotImplemented
+        r = rotation.getRotation()
+        return Vector(*r.apply(self.coordinates))
+
+    @vectorOperator
+    def cartesianToSpherical(self):
+        """Returns this vector in spherical coordinates (rho, theta, phi)"""
+        rho = math.hypot(self.x, self.y, self.z)
+        theta = math.atan2(self.y, self.x) - math.pi/2
+        phi = math.atan2(self.z, math.hypot(self.x,self.y))
+        return Vector(rho, theta, phi)
+
+    @zeroPreservingVectorOperator
+    def rotatedBy(self, angleOrOrientation) -> Vector:
+        """Return a vector equal to this one rotated counterclockwise by the given angle."""
+        if isinstance(angleOrOrientation, Orientation):
+            return self.applyRotation(angleOrOrientation)
+        x, y, z = self.x, self.y, self.z
+        c, s = cos(angleOrOrientation), sin(angleOrOrientation)
+        return Vector((c * x) - (s * y), (s * x) + (c * y), z)
+
+    @vectorOperator
+    def offsetRotated(self, angleOrOrientation, offset) -> Vector:
+        ro = offset.rotatedBy(angleOrOrientation)
+        return self + ro
+
+    @vectorOperator
+    def offsetRadially(self, radius, heading) -> Vector:
+        return self.offsetRotated(heading, Vector(0, radius))
+
+    @scalarOperator
+    def distanceTo(self, other) -> float:
+        if not isinstance(other, Vector):
+            return other.distanceTo(self)
+        dx, dy, dz = other.toVector() - self
+        return math.hypot(dx, dy, dz)
+
+    @scalarOperator
+    def angleTo(self, other) -> float:
+        return self.azimuthTo(other)
+
+    @scalarOperator
+    def azimuthTo(self, other) -> float:
+        dx, dy, dz = other.toVector() - self
+        return normalizeAngle(math.atan2(dy, dx) - (math.pi / 2))
+
+    @scalarOperator
+    def altitudeTo(self, other) -> float:
+        dx, dy, dz = other.toVector() - self
+        return normalizeAngle(math.atan2(dz, math.sqrt(dx**2 + dy**2)))
+
+    @scalarOperator
+    def angleWith(self, other) -> float:
+        """Compute the signed angle between self and other.
+
+        The angle is positive if other is counterclockwise of self (considering
+        the smallest possible rotation to align them).
+        """
+        x, y = self.x, self.y
+        ox, oy = other.x, other.y
+        return normalizeAngle(math.atan2(oy, ox) - math.atan2(y, x))
+
+    @scalarOperator
+    def norm(self) -> float:
+        return hypot(*self.coordinates)
+
+    @scalarOperator
+    def dot(self, other) -> float:
+        x, y, z = self.x, self.y, self.z
+        ox, oy, oz = other.x, other.y, other.z
+        return (x * ox) + (y * oy) + (z * oz)
+
+    @vectorOperator
+    def cross(self, other) -> Vector:
```

Comment:
This is never used. If you think it might come in handy, we could keep it but should make it actually return a `Vector`.